PolyEngine/Engine/Src/DebugDrawSystem.cpp

```diff
@@ -0,0 +1,236 @@
+				result.Y = vec.Y;
+			if (comp(vec.Z, result.Z))
+				result.Z = vec.Z;
+		}
+
+		return result;
+	}
+
+	/// <summary>Produces two orthonormal vectors from a single normal vector.</summary>
+	/// Building an Orthonormal Basis, Revisited; http://jcgt.org/published/0006/01/01/
+	void branchlessONB(const Vector & n, Vector & b1, Vector & b2)
+	{
+		float sign = std::copysignf(1.0f, n.Z);
+		const float a = -1.0f / (sign + n.Z);
+		const float b = n.X * n.Y * a;
+		b1 = Vector(1.0f + sign * n.X * n.X * a, sign * b, -sign * n.X);
+		b2 = Vector(b, sign + n.Y * n.Y * a, -n.Y);
+	}
+}
+
+void DebugDrawSystem::DebugRenderingUpdatePhase(World* world)
+{
+	gDebugConfig.DebugRender = false;
+	for (auto& kv : world->GetWorldComponent<ViewportWorldComponent>()->GetViewports())
+	{
+		CameraComponent* cameraCmp = kv.second.GetCamera();
+		if (cameraCmp->GetRenderingMode() == eRenderingModeType::DEBUG)
+			gDebugConfig.DebugRender = true;
+	}
+
+	if(!gDebugConfig.DebugRender)
+		return;
+
+	enum class RenderMode : int { /*POINT = 1,*/ LINE = 1, /*STRING,*/ _COUNT };
+
+	// iterate RenderMode::_COUNT times to create shapes defining debug primitives
+	for(int renderMode = static_cast<int>(RenderMode::LINE); renderMode < static_cast<int>(RenderMode::_COUNT); ++renderMode)
+	{
+		for(auto componentsTuple : world->IterateComponents<DebugDrawableComponent, MeshRenderingComponent, TransformComponent>())
+		{
+			const auto ddrawCmp = std::get<DebugDrawableComponent*>(componentsTuple);
+			const auto meshCmp = std::get<MeshRenderingComponent*>(componentsTuple);
+			const auto transformCmp = std::get<TransformComponent*>(componentsTuple);
+
+			if (meshCmp == nullptr)
+				continue;
+
+			if (!gDebugConfig.DebugDrawPresets.IsSet(ddrawCmp->entityPreset))
+				continue;
+
+			Vector objPos = transformCmp->GetGlobalTranslation();
+
+			const Matrix& objTransform = transformCmp->GetGlobalTransformationMatrix();
+
+			// spawn a box for every mesh, in correct size
+			for(const auto subMesh : meshCmp->GetMesh()->GetSubMeshes())
+			{
+				const auto & meshVerticesPositions = subMesh->GetMeshData().GetPositions();
+
+				// find mins and maxs of each mesh coordinates
+				// in order to create AABB
+				auto minMeshVector = Util::findExtremum(meshVerticesPositions, std::less<float>());
+				auto maxMeshVector = Util::findExtremum(meshVerticesPositions, std::greater<float>());
+
+				Vector minVector(minMeshVector.X, minMeshVector.Y, minMeshVector.Z);
+				Vector maxVector(maxMeshVector.X, maxMeshVector.Y, maxMeshVector.Z);
+
+				minVector = objTransform * minVector;
+				maxVector = objTransform * maxVector;
+
+				const auto boundingOffset = Vector(0.04f, 0.04f, 0.04f);
+				// move vector away a little bit from a mesh
+				minVector -= boundingOffset;
+				maxVector += boundingOffset;
+
+				EmitBox(world, minVector, maxVector);
+			}
+		}
+
+		for (auto componentsTuple : world->IterateComponents<RigidBody2DComponent, DebugDrawableComponent, TransformComponent>())
```

Comment:
I'm not sure if this is DebugDraw system job to know about RigidBody2D and how to draw velocity for it. We probably need separate PhysicsDebug system. Mark it as @fixme.